### DISCUSSION TITLE
STM32L4 Power management help

### DISCUSSION THREAD

**Discussion by darexa** [zephyrproject-rtos/zephyr#43530] [Q&A] [answered]
So I've started looking at power management in my application.

What I've done:

CONFIG_PM=y

Power states added to my board:
`
	power-states {
		stop0: state0 {
			compatible = "zephyr,power-state";
			power-state-name = "suspend-to-idle";
			substate-id = <1>;
			min-residency-us = <500>;
		};
		stop1: state1 {
			compatible = "zephyr,power-state";
			power-state-name = "suspend-to-idle";
			substate-id = <2>;
			min-residency-us = <700>;
		};
		stop2: state2 {
			compatible = "zephyr,power-state";
			power-state-name = "suspend-to-idle";
			substate-id = <3>;
			min-residency-us = <1000>;
		};
	};
`
Currently just testing things out with STOP0:

`
&cpu0 {
	cpu-power-states = <&stop0>;
};
`

I've commented out pretty much everything from my main thread, I'm only sleeping for 5 seconds at a time.

Running without PM, consumption is around 8mA.
Running with PM, consumption is around 3mA.
Running without PM and entering STOP0 manually is around 1mA.

I was expecting the last two steps to give identical results but they're not and I don't understand why,¨
I've attached callbacks to 
`pm_notifier_register(&pm_callbacks);`
to try to figure something out and it seems low power mode is entered and exited constantly so I'm guessing that may be the reason.
No clue why it would do that though.
Anyone worked with PM on STM32L4 and have ideas?
Thanks

**Comment by darexa** [comment]
Haven't solved thus yet, but found that the ticks that come into the policy is always around ~4000. My ticks per second is set to 4096 so seems like it schedules to sleep for about one second each time. From debug prints I can see it actually schedules that like every 100ms so it must wake for some reason.

**Comment by erwango** [comment, marked as answer]
STM32 power management in Zephyr relies on lptim being used as tick source when PM=y.
Did you enabled lptim node in you board dts ?

**Reply by darexa** [reply]
Nope, thanks. You just made my week! It's actually going to the correct power level now. Haven't tried it in my full app yet but this is a giant step forward.
Thanks again!

**Comment by darexa** [comment]
@erwango Sorry to bother you again but you seem to know much about STM and Zephyr. 
I'm trying to set, for example, lpuart1 as a wakeup source.
I have added `wakeup-source;` to the device tree for lpuart1.
Upon boot I then check `pm_device_wakeup_is_capable(lpuart1)` and it returns false.
Do you know if I need to enable anything further? PM and PM_DEVICE is enabled at the moment.

**Reply by erwango** [reply]
@darexa Well, this part is not yet fully available.
I'm not sure why `pm_device_wakeup_is_capable(lpuart1)` returns false, but for sure stm32 uart driver doesn't allow yet to configure a device as wake up source. We're working on it, but it will take some time.

**Reply by darexa** [reply]
@erwango I see, thank you. I will see if I can work around it.
